src/runtime/FunctionObject.cpp

```diff
@@ -320,6 +320,131 @@ Object* FunctionObject::newInstance(ExecutionState& state, const size_t argc, Va
+        CallNativeFunctionData* code = m_codeBlock->nativeFunctionData();
+        FunctionEnvironmentRecordSimple record(this);
+        LexicalEnvironment env(&record, outerEnvironment());
+        ExecutionContext ec(ctx, state.executionContext(), &env, isStrict);
+
+        size_t len = m_codeBlock->parameterCount();
+        if (argc < len) {
+            Value* newArgv = (Value*)alloca(sizeof(Value) * len);
+            for (size_t i = 0; i < argc; i++) {
+                newArgv[i] = argv[i];
+            }
+            for (size_t i = argc; i < len; i++) {
+                newArgv[i] = Value();
+            }
+            argv = newArgv;
+        }
+
+        Value receiver = receiverSrc;
+        // prepare receiver
+        if (UNLIKELY(isStrict == false)) {
+            if (receiver.isUndefinedOrNull()) {
+                receiver = ctx->globalObject();
+            } else {
+                receiver = receiver.toObject(state);
+            }
+        }
+
+        record.bindThisValue(state, receiver);
+        if (receiverSrc.isObject()) {
+            record.setNewTarget(receiverSrc.asObject());
+        }
+
+        ExecutionState newState(ctx, &state, &ec, &receiver);
+
+        try {
+            return code->m_fn(newState, receiver, argc, argv, isNewExpression);
+        } catch (const Value& v) {
+            ByteCodeInterpreter::processException(newState, v, &ec, SIZE_MAX);
+        }
+    }
+
+    // prepare ByteCodeBlock if needed
+    if (UNLIKELY(m_codeBlock->asInterpretedCodeBlock()->byteCodeBlock() == nullptr)) {
+        generateBytecodeBlock(state);
+    }
+
+    ByteCodeBlock* blk = m_codeBlock->asInterpretedCodeBlock()->byteCodeBlock();
+
+    size_t registerSize = blk->m_requiredRegisterFileSizeInValueSize;
+    size_t stackStorageSize = m_codeBlock->asInterpretedCodeBlock()->identifierOnStackCount();
+    size_t literalStorageSize = blk->m_numeralLiteralData.size();
+    size_t parameterCopySize = std::min(argc, (size_t)m_codeBlock->parameterCount());
+
+    // prepare env, ec
+    ASSERT(m_codeBlock->canAllocateEnvironmentOnStack() == true);
+    FunctionEnvironmentRecordSimple record(this);
+    LexicalEnvironment env(&record, outerEnvironment());
+    ExecutionContext ec(ctx, state.executionContext(), &env, isStrict);
+
+    if (receiverSrc.isObject() == true) {
+        record.setNewTarget(receiverSrc.asObject());
+    }
+
+    Value* registerFile = (Value*)alloca((registerSize + stackStorageSize + literalStorageSize) * sizeof(Value));
+    Value* stackStorage = registerFile + registerSize;
+    Value* literalStorage = stackStorage + stackStorageSize;
+    memcpy(literalStorage, blk->m_numeralLiteralData.data(), literalStorageSize * sizeof(Value));
+
+    // prepare receiver
+    ASSERT(m_codeBlock->isArrowFunctionExpression() == false);
+    ASSERT(thisMode() != ThisMode::Lexical);
+
+    if (isStrict == false) {
+        if (receiverSrc.isUndefinedOrNull()) {
```

Comment:
For readability / State expression explicitly
if (receiver.isUndefinedOrNull() == true)